File: app/models/users/UserModels.scala
package models.users

import models.common.Email
import models.meta.{EntityTypes, EntityType}
import models.{EntityId, Entity}
import models.auth.AuthAccount
import org.joda.time.DateTime
import play.api.libs.json.{OFormat, Format, Json}
import serializers.IdSerializer
import util.json.PlayJson

case class Name(first: String, last: String)

object Name {

  implicit val format: Format[Name] = Json.format[Name]
}

case class User(
  id: UserId,
  username: String,
  email: Email,
  name: Name,
  auth: Seq[AuthAccount] = Seq.empty,
  created: DateTime = DateTime.now()
) extends Entity

object User {

  implicit val format: Format[User] = Json.format[User]
}

case class UserId(value: String) extends AnyVal with EntityId {

  override def entityType: EntityType = EntityTypes.User
}

object UserId extends IdSerializer[UserId]

File: public/typescript/app/models/user-models.ts
/// <reference path="../reference.ts" />

interface Name {
  first: string
  last: string
}

class AuthAccount extends Entity {
}

class User extends Entity {
  constructor(
    id: string,
    public username: string,
    public name: Name,
    public email: Email,
    public auth: AuthAccount[] = []) {
    super(id)
  }
}
